### PR TITLE
Modernize C++

### PR DESCRIPTION
**Pull request by marcusmueller**
# Pull Request Details

## Description

Three things:

* using range-based for loops, especially for iterating through tags (this is a pattern I'd like to establish)
* using `string == "value"` instead of `string.compare("value")` (guess this is a C-ism)
* consistent `override`
* assert.h -> cassert
* ruthlessness and surprise

Seemed to be a good time to do this, what with kicking of a "fresh" master

## Which blocks/areas does this affect?

all over the tree - motivated by the string comparisons in gr-iio.

## Testing Done

Testsuite passes. No new tests written.

## Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/master/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/master/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [x] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

**Comment by willcode** [comment]
Would be good to backport to 3.9. Much of the patch applies as is, some tweaking will be needed.

File: gnuradio-runtime/lib/math/qa_fxpt_vco.cc [hunk 1]
```diff
@@ -38,15 +38,15 @@ BOOST_AUTO_TEST_CASE(t0)
         input[i] = sin(double(i));
     }
 
-    for (int i = 0; i < SIN_COS_BLOCK_SIZE; i++) {
+    for (float i : input) {
```
Comment (ThomasHabets) [inline gnuradio-runtime/lib/math/qa_fxpt_vco.cc:41]: Pretty poor name for the float, now.

Maybe just `in`?

File: gnuradio-runtime/lib/top_block_impl.cc [hunk 2]
```diff
@@ -50,10 +50,9 @@ make_scheduler(flat_flowgraph_sptr ffg, int max_noutput_items, bool catch_except
         if (!v)
             factory = scheduler_table[0].f; // use default
         else {
-            for (size_t i = 0; i < sizeof(scheduler_table) / sizeof(scheduler_table[0]);
-                 i++) {
-                if (strcmp(v, scheduler_table[i].name) == 0) {
-                    factory = scheduler_table[i].f;
+            for (auto& i : scheduler_table) {
```
Comment (ThomasHabets) [inline gnuradio-runtime/lib/top_block_impl.cc:53]: Not a great name here either. But more generally: Many (most, probably) of these should live with getting a `const` for readability.

File: gr-blocks/lib/rotator_cc_impl.h [hunk 3]
```diff
@@ -23,7 +23,7 @@ struct phase_inc_update_t {
     double phase_inc;
 };
 
-bool cmp_phase_inc_update_offset(phase_inc_update_t lhs, phase_inc_update_t rhs)
+inline bool cmp_phase_inc_update_offset(phase_inc_update_t lhs, phase_inc_update_t rhs)
```
Comment (ThomasHabets) [inline gr-blocks/lib/rotator_cc_impl.h:26]: Unrelated?

File: gr-filter/lib/qa_firdes.cc [hunk 4]
```diff
@@ -21,8 +21,8 @@ using std::vector;
 void print_taps(std::ostream& s, vector<float>& v)
 {
     std::streamsize tmp = s.precision(9);
-    for (unsigned int i = 0; i < v.size(); i++) {
-        s << v[i] << ", ";
+    for (float i : v) {
```
Comment (ThomasHabets) [inline gr-filter/lib/qa_firdes.cc:24]: Again a float named i.

File: gr-iio/lib/attr_sink_impl.h [hunk 5]
```diff
@@ -40,7 +40,7 @@ class attr_sink_impl : public attr_sink
                    const std::string& channel,
                    attr_type_t type,
                    bool output);
-    ~attr_sink_impl();
+    ~attr_sink_impl() override;
```
Comment (ThomasHabets) [inline gr-iio/lib/attr_sink_impl.h:43]: Should be separate PR?

Comment (ThomasHabets) [reply gr-iio/lib/attr_sink_impl.h:43]: (also other instances)

Comment (ThomasHabets) [reply gr-iio/lib/attr_sink_impl.h:43]: Never mind, same PR different commit is good.

File: gnuradio-runtime/lib/math/qa_fxpt_vco.cc (hunk 1 above)
Comment (marcusmueller) [reply gnuradio-runtime/lib/math/qa_fxpt_vco.cc:41]: agreed. This is just the autoconversion that clang-tidy does. When the original variable was short and the container ends in an "s", the clang-tidy pass tries to guess a singular and uses that (eg `for(int i = 0; i<inputs.size(); ++i)`->`for(auto& input : inputs)`), but here there's no plural-container :)

File: gnuradio-runtime/lib/top_block_impl.cc (hunk 2 above)
Comment (marcusmueller) [reply gnuradio-runtime/lib/top_block_impl.cc:53]: yeah the heuristic with which the `const`ness is determined is really rather ... conservative. At least it doesn't do any copies as far as I looked through it.

File: gr-filter/lib/qa_firdes.cc (hunk 4 above)
Comment (marcusmueller) [reply gr-filter/lib/qa_firdes.cc:24]: yeah, will fix.

**Comment by marcusmueller** [comment]
@ThomasHabets fixed the readability issues where arisen from range-based for-conversion; drive-by modernized other things, as well.

File: gr-dtv/lib/dvb/dvb_bbscrambler_bb_impl.cc [hunk 6]
```diff
@@ -31,210 +32,12 @@ dvb_bbscrambler_bb_impl::dvb_bbscrambler_bb_impl(dvb_standard_t standard,
                                                  dvb_code_rate_t rate)
     : gr::sync_block("dvb_bbscrambler_bb",
                      gr::io_signature::make(1, 1, sizeof(unsigned char)),
-                     gr::io_signature::make(1, 1, sizeof(unsigned char)))
+                     gr::io_signature::make(1, 1, sizeof(unsigned char))),
+      kbch(dvb::constants::kbch_mapping.at(framesize).at(rate)),
```
Comment (marcusmueller) [inline gr-dtv/lib/dvb/dvb_bbscrambler_bb_impl.cc:36]: now we just need to do that everywhere and we get very compact source files :D

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc [hunk 7]
```diff
@@ -27,42 +28,46 @@ atsc_equalizer::sptr atsc_equalizer::make()
     return gnuradio::make_block_sptr<atsc_equalizer_impl>();
 }
 
-static float bin_map(int bit) { return bit ? +5 : -5; }
+constexpr float bin_map(unsigned char bit) { return bit ? +5 : -5; }
 
-static void init_field_sync_common(float* p, int mask)
+template <size_t N>
+constexpr std::array<float, N> init_field_sync_common(bool invert)
```
Comment (ThomasHabets) [inline gr-dtv/lib/atsc/atsc_equalizer_impl.cc:34]: `std::array` does not have move assignment/constructor, thus every call here will cause `N*4` bytes to be copied (unless RVO). Is that the intention?

**Review by ThomasHabets — COMMENTED** [review]
Only a couple of outstdanding comments.

File: gr-digital/lib/ofdm_carrier_allocator_cvc_impl.cc [hunk 8]
```diff
@@ -143,8 +143,8 @@ int ofdm_carrier_allocator_cvc_impl::work(int noutput_items,
 
     memset((void*)out, 0x00, sizeof(gr_complex) * d_fft_len * noutput_items);
     // Copy Sync word
-    for (unsigned i = 0; i < d_sync_words.size(); i++) {
-        memcpy((void*)out, (void*)&d_sync_words[i][0], sizeof(gr_complex) * d_fft_len);
+    for (const auto& sync_word : d_sync_words) {
+        memcpy((void*)out, (void*)&sync_word[0], sizeof(gr_complex) * d_fft_len);
```
[the remark this answers is not in the document]
Comment (marcusmueller) [reply gr-digital/lib/ofdm_carrier_allocator_cvc_impl.cc:147]: true! Intuitively wanted to opt for `std::copy_n`, as we have the FFT length, and my *guess* is that a contiguous memory copy is more efficient than having to check for `!= end` after every item. However, `std::copy` has a specialization that seems to call memcpy under the hood; and it seems to kick on current compilers.

File: gr-dtv/lib/atsc/atsc_equalizer_impl.cc (hunk 7 above)
Comment (ThomasHabets) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc:34]: `std::array` is stack (well… technically "auto storage duration") allocated (including the data), so cannot even in theory do a pointer swap for move. It can do per-element move of all its elements, but for floats that's not an optimization.

Your godbolt example shows two differences:
1. it likely triggers RVO, which always means no copy nor move. (I don't think it's [mandatory RVO](https://en.cppreference.com/w/cpp/language/copy_elision))
2. the function is inlined, so not really returned.

[Adding `-fno-inline` takes care of the second, and making RVO impossible](https://gcc.godbolt.org/z/6GzGcn9eo) (for example by calling a second time) causes a call to memcpy (or [in this example](https://gcc.godbolt.org/z/61n6x7rfe) a `rep movsq`).

So if the only correct way to call this function is in the mandatory RVO situations? That may be a bit of a foot gun for non-initialization at caller site. And I don't think as-is this is mandatory RVO since it matches the non-mandatory's section "In a return statement, when the operand is the name of a non-volatile object with automatic storage duration".

I guess since this is constexpr it may actually be mandatory RVO, but still only applies when caller makes it possible (in initialization).

As for `std::move`, despite its name it doesn't move. It just converts lvalue to rvalue, and you can copy an rvalue. But also yes technically `std::array` can be moved, but it's just an elementwise move, which for array of float means a de facto copy.

Also examples like [this](https://gcc.godbolt.org/z/fq8YovMed) show that `std::move()` doesn't force a move, only enables it.

But yes, technically I was wrong by saying `std::array` doesn't "have" a move assign/constructor. It does, it just doesn't do the expected thing, because it can't.

Comment (marcusmueller) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc:34]: Oh, wow! First of all: thanks for letting me occupy you so much! I'm thankful, and I'm learning a lot :)

So, the right approach here is instead letting `init_field_sync_common` return a `std::vector` value, making `training_sequence1,2` std::vectors and using `std::vector`s (actually moving) move c'tor in through `training_sequence1(std::move(init_field_sync_common<…>(false))` in the initializer list?

Comment (ThomasHabets) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc:34]: Maybe. But at the cost of one indirection when the sequences are accessed.

I'd say there are three options:

1. As this PR currently does: Depending on compiler and options could do a needless copy, but the training data is inlined with the impl object so no indirection.
2. Switch to `std::vector`. No copies on init, but an indirection on access to the data, as `vector`s don't have a "small string optimization".
3. Call init functions, as before, but now with `std::array` instead of C arrays. No performance cost, as init is done in-place, and data inlined. Only downside is that it's a shame to not use the neatness of initializer list.

I've not played with dtv, but I suspect the important bit is to make the data path fast, which means (1) or (3). I'd lean towards (3).

Comment (marcusmueller) [reply gr-dtv/lib/atsc/atsc_equalizer_impl.cc:34]: opted for (3)!

File: gr-digital/lib/constellation.cc [hunk 9]
```diff
@@ -353,21 +355,16 @@ std::vector<float> constellation::soft_decision_maker(gr_complex sample)
...
-            d_im_max = d_constellation[i].imag();
-
-        if (d_constellation[i].real() < d_re_min)
-            d_re_min = d_constellation[i].real();
-        if (d_constellation[i].imag() < d_im_min)
-            d_im_min = d_constellation[i].imag();
+    // Find min/max of constellation for both real and imag, axes.
+    d_re_min = std::numeric_limits<decltype(d_re_min)>::max();
+    d_im_min = std::numeric_limits<decltype(d_re_min)>::max();
+    d_re_max = -d_re_min;
```
Comment (ThomasHabets) [inline gr-digital/lib/constellation.cc:361]: I think it's safer and more correct to use `std::numeric_limits<decltype(d_re_max)>::lowest()`.

File: gr-digital/lib/ofdm_carrier_allocator_cvc_impl.cc [hunk 10]
```diff
@@ -137,36 +134,38 @@ int ofdm_carrier_allocator_cvc_impl::work(int noutput_items,
...
+    auto input_cnt = static_cast<std::size_t>(ninput_items[0]);
+    auto end_pointer = out + d_fft_len * noutput_items;
     std::vector<tag_t> tags;
 
-    memset((void*)out, 0x00, sizeof(gr_complex) * d_fft_len * noutput_items);
     // Copy Sync word
-    for (unsigned i = 0; i < d_sync_words.size(); i++) {
-        memcpy((void*)out, (void*)&d_sync_words[i][0], sizeof(gr_complex) * d_fft_len);
+    for (const auto& sync_word : d_sync_words) {
+        std::copy(sync_word.cbegin(), sync_word.end(), out);
```
Comment (ThomasHabets) [inline gr-digital/lib/ofdm_carrier_allocator_cvc_impl.cc:145]: Looks a bit odd to do `.cbegin()` with `.end()`. `.cend()`?

**Review by ThomasHabets — APPROVED** [review]
Aside from those two things: Looks good

File: gnuradio-runtime/lib/flat_flowgraph.cc [hunk 11]
```diff
@@ -315,12 +313,12 @@ void flat_flowgraph::merge_connections(flat_flowgraph_sptr old_ffg)
     }
 
     // Now connect inputs to outputs, reusing old buffer readers if they exist
-    for (basic_block_viter_t p = d_blocks.begin(); p != d_blocks.end(); p++) {
-        block_sptr block = cast_to_block_sptr(*p);
+    for (auto& d_block : d_blocks) {
```
Comment (mbr0wn) [inline gnuradio-runtime/lib/flat_flowgraph.cc:313]: The loop variable should not start with `d_` as per our coding guidelines.

File: gnuradio-runtime/lib/flat_flowgraph.cc [hunk 12]
```diff
@@ -413,27 +411,27 @@ void flat_flowgraph::setup_buffer_alignment(block_sptr block)
 std::string flat_flowgraph::edge_list()
 {
     std::stringstream s;
-    for (edge_viter_t e = d_edges.begin(); e != d_edges.end(); e++)
-        s << (*e) << std::endl;
+    for (auto& d_edge : d_edges)
```
Comment (mbr0wn) [inline gnuradio-runtime/lib/flat_flowgraph.cc:411]: Same comment (`d_`)

**Comment by mbr0wn** [comment]
> ruthlessness and surprise

Surprises... aren't a great thing to have in a PR.

File: gnuradio-runtime/include/gnuradio/block.h [hunk 13]
```diff
@@ -545,9 +545,9 @@ class GR_RUNTIME_API block : public basic_block
      * \brief Allocate the block_detail and necessary output buffers for this
      * block.
      */
-    void allocate_detail(int ninputs,
-                         int noutputs,
-                         const std::vector<int>& downstream_max_nitems_vec,
+    void allocate_detail(unsigned int ninputs,
```
Comment (mbr0wn) [inline gnuradio-runtime/include/gnuradio/block.h:548]: `unsigned int` often bothers me, reminds me of C of yore. Wouldn't `size_t` be the more appropriate type?

File: gnuradio-runtime/include/gnuradio/block.h [hunk 14]
```diff
@@ -545,9 +545,9 @@ class GR_RUNTIME_API block : public basic_block
...
+                         unsigned int noutputs,
+                         const std::vector<unsigned int>& downstream_max_nitems_vec,
```
Comment (mbr0wn) [inline gnuradio-runtime/include/gnuradio/block.h:550]: Here, for example, I know it's preference, but I don't like that space in the template param.

**Review by mbr0wn — COMMENTED** [review]
Can you maybe split this up? Some of these commits aren't even truly modernizing (like the extra inlining). I'm very glad people like you and @ThomasHabets take on the work of de-crufting and modernizing our codebase, but this is pretty big and might be easier to review and merge if you put up multiple PRs.

**Comment by mormj** [comment]
Love the changes, but agree with @mbr0wn on splitting this up - many good changes, but it is tough to review.  Also concerned that the extent of the changes would be questionable to backport - and not backporting would cause an unnecessary code diversion

Closing for now - I think smaller PRs would help alleviate the concerns
